### ISSUE TITLE
ANMN_NRS_BGC and SOOP_AUSCPR intermittent timeout issues

### ISSUE THREAD

**Issue by jonescc** [aodn/harvesters#271]
ANMN_NRS_BGC and SOOP_AUSCPR both fall over occasionally with timeout errors trying to connect to the CMAR server.  The issue is with the CMAR server occasionally taking too long to respond to WFS requests.

**Comment by jonescc** [comment]
Note that the timeout can be set in the advanced properties of the sWFSInput.   Its currently set to 3000 ms in each of these harvesters.

**Comment by mhidas** [comment]
@jonescc Thanks. I will increase the timeout values. At the moment the CMAR geoserver is taking 3-4minutes to respond, so I'll just set it to 10min for each request. I guess the harvester has time to be patient.

I have also pointed this out to the people responsible for that server in case they can do anything to improve it.

**Comment by mhidas** [comment]
Ok, CMAR appear to have fixed this issue. The final word from their IT guy was:
> I think we've worked out the issue, seems there's an IPv6 configuration problem with our server which our support staff have remedied.

However, it still occasionally takes 10 or so seconds to respond, so I'll need to increase that timeout threshold in the harvesters anyway. One minute should be enough I guess.

**Comment by mhidas** [comment]
Getting timeouts again on ANMN_NRS_BGC. The timeout limit is now set to 60sec. Maybe it needs longer?

The exception messages vary a bit, in case that is significant:

On 2015-09-02 19:16:41+10:00
```
Exception in component sWfsInput_1_GI
java.net.SocketTimeoutException: connect timed out
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:339)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:200)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:182)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:579)
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:618)
        at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:275)
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:371)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:932)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1300)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
        at org.geotools.data.wfs.protocol.http.SimpleHttpProtocol$SimpleHttpResponse.getResponseStream(SimpleHttpProtocol.java:70)
        at org.geotools.data.wfs.WFSDataStoreFactory.loadCapabilities(WFSDataStoreFactory.java:779)
        at org.geotools.data.wfs.WFSDataStoreFactory.createDataStore(WFSDataStoreFactory.java:412)
        at anmn_nrs_bgc.station_harvest_0_1.station_harvest.sWfsInput_1_GIProcess(station_harvest.java:2236)
        at anmn_nrs_bgc.station_harvest_0_1.station_harvest.tPostgresqlConnection_1Process(station_harvest.java:657)
        at anmn_nrs_bgc.station_harvest_0_1.station_harvest.iIncludeSdiLibraries_1Process(station_harvest.java:539)
        at anmn_nrs_bgc.station_harvest_0_1.station_harvest.runJobInTOS(station_harvest.java:4538)
        at anmn_nrs_bgc.station_harvest_0_1.station_harvest.runJob(station_harvest.java:4395)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tRunJob_1Process(ANMN_NRS_BGC_harvester.java:2539)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tPostgresqlClose_1Process(ANMN_NRS_BGC_harvester.java:2400)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.iPostgresqlDbUpdate_1Process(ANMN_NRS_BGC_harvester.java:2297)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tPostgresqlConnection_1Process(ANMN_NRS_BGC_harvester.java:1551)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.iIncludeSdiLibraries_1Process(ANMN_NRS_BGC_harvester.java:1417)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.runJobInTOS(ANMN_NRS_BGC_harvester.java:5769)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.main(ANMN_NRS_BGC_harvester.java:5529)
```

On 2015-09-03 19:21:17+10:00:
```
Exception in component sWfsInput_1_GI
java.util.NoSuchElementException: Could not aquire feature:java.net.SocketTimeoutException: Read timed out
        at org.geotools.data.store.NoContentIterator.next(NoContentIterator.java:58)
        at org.geotools.data.store.NoContentIterator.next(NoContentIterator.java:43)
        at org.geotools.feature.collection.DelegateFeatureIterator.next(DelegateFeatureIterator.java:58)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.sWfsInput_1_GIProcess(plankton_zooplankton_harvest.java:3268)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.tPostgresqlConnection_1Process(plankton_zooplankton_harvest.java:778)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.iIncludeSdiLibraries_1Process(plankton_zooplankton_harvest.java:660)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.runJobInTOS(plankton_zooplankton_harvest.java:7086)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.runJob(plankton_zooplankton_harvest.java:6915)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tRunJob_11Process(ANMN_NRS_BGC_harvester.java:3585)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tRunJob_1Process(ANMN_NRS_BGC_harvester.java:2580)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tPostgresqlClose_1Process(ANMN_NRS_BGC_harvester.java:2400)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.iPostgresqlDbUpdate_1Process(ANMN_NRS_BGC_harvester.java:2297)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.tPostgresqlConnection_1Process(ANMN_NRS_BGC_harvester.java:1551)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.iIncludeSdiLibraries_1Process(ANMN_NRS_BGC_harvester.java:1417)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.runJobInTOS(ANMN_NRS_BGC_harvester.java:5769)
        at anmn_nrs_bgc.anmn_nrs_bgc_harvester_0_1.ANMN_NRS_BGC_harvester.main(ANMN_NRS_BGC_harvester.java:5529)
Caused by: java.net.SocketTimeoutException: Read timed out
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at sun.net.www.protocol.http.HttpURLConnection$6.run(HttpURLConnection.java:1675)
        at sun.net.www.protocol.http.HttpURLConnection$6.run(HttpURLConnection.java:1673)
        at java.security.AccessController.doPrivileged(Native Method)
        at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1671)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1244)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
        at org.geotools.data.wfs.protocol.http.SimpleHttpProtocol$SimpleHttpResponse.getResponseStream(SimpleHttpProtocol.java:70)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_Protocol.issueGetRequest(WFS_1_1_0_Protocol.java:642)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_Protocol.issueGetFeatureGET(WFS_1_1_0_Protocol.java:476)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_DataStore.sendGetFeatures(WFS_1_1_0_DataStore.java:400)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_DataStore.executeGetFeatures(WFS_1_1_0_DataStore.java:304)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_DataStore.getFeatureReader(WFS_1_1_0_DataStore.java:330)
        at org.geotools.data.wfs.v1_1_0.WFSFeatureCollection.reader(WFSFeatureCollection.java:167)
        at org.geotools.data.store.DataFeatureCollection.openIterator(DataFeatureCollection.java:230)
        at org.geotools.data.store.DataFeatureCollection.iterator(DataFeatureCollection.java:198)
        at org.geotools.data.store.DataFeatureCollection.features(DataFeatureCollection.java:187)
        at org.geotools.data.store.DataFeatureCollection.features(DataFeatureCollection.java:78)
        at anmn_nrs_bgc.plankton_zooplankton_harvest_0_1.plankton_zooplankton_harvest.sWfsInput_1_GIProcess(plankton_zooplankton_harvest.java:3266)
        ... 12 more
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.read(SocketInputStream.java:152)
        at java.net.SocketInputStream.read(SocketInputStream.java:122)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:442)
        at sun.security.ssl.InputRecord.read(InputRecord.java:480)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:927)
        at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:884)
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:102)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
        at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:687)
        at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:633)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1323)
        at sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:2678)
        at java.net.URLConnection.getContentType(URLConnection.java:522)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getContentType(HttpsURLConnectionImpl.java:406)
        at org.geotools.data.wfs.protocol.http.SimpleHttpProtocol$SimpleHttpResponse.getContentType(SimpleHttpProtocol.java:55)
        at org.geotools.data.wfs.v1_1_0.WFS_1_1_0_Protocol.issueGetRequest(WFS_1_1_0_Protocol.java:641)
        ... 22 more
```

**Comment by mhidas** [comment]
I just reran the ANMN_NRS_BGC harvester manually and it succeeded.

**Comment by lbesnard** [comment]
@mhidas , I'm looking at the Jenkins talend pipeline, trying to clean things and deploy things which should have been deployed. I can see that the SOOP AUSCPR was modified on github, but the changes were not deployed nor to RC nor to PROD. Is it supposed to be deployed ?

**Comment by mhidas** [comment]
@lbesnard I'm a bit puzzled, because as far as I can tell from github, the last change made to SOOP AUSCPR was merged into master in #272 and has been deployed. Anyone else know of any changes?

Anyway, go ahead, deploy it to RC and see what happens.

**Comment by mhidas** [comment]
@lbesnard Where did we get to with this? Did you re-deploy AUSCPR?

Anyway, both AUSCPR and NRS_BGC have been running without issues for a couple of weeks now, so as far as I'm concerned we can close this issue.
